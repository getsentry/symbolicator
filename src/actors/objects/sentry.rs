use std::sync::Arc;
use std::time::Duration;

use actix_web::{client, HttpMessage};

use failure::Fail;
use futures::future::{self, Either};
use futures::{Future, IntoFuture, Stream};
use tokio_retry::strategy::{jitter, ExponentialBackoff};
use tokio_retry::Retry;
use tokio_threadpool::ThreadPool;

use crate::actors::common::cache::Cacher;
use crate::actors::objects::{
    DownloadStream, FetchFileRequest, FileId, ObjectError, ObjectErrorKind, PrioritizedDownloads,
    SentryFileId, USER_AGENT,
};
use crate::sentry::SentryFutureExt;
use crate::types::{ArcFail, FileType, ObjectId, Scope, SentrySourceConfig};

#[derive(Debug, Fail, Clone, Copy)]
pub enum SentryErrorKind {
    #[fail(display = "failed parsing JSON response from Sentry")]
    Parsing,

    #[fail(display = "bad status code from Sentry")]
    BadStatusCode,

    #[fail(display = "failed sending request to Sentry")]
    SendRequest,
}

#[derive(serde::Deserialize)]
struct FileEntry {
    id: String,
    // TODO: Add more fields
}

symbolic::common::derive_failure!(
    SentryError,
    SentryErrorKind,
    doc = "Errors happening while fetching data from Sentry"
);

pub fn prepare_downloads(
    source: &Arc<SentrySourceConfig>,
    scope: Scope,
    _filetypes: &'static [FileType],
    object_id: &ObjectId,
    threadpool: Arc<ThreadPool>,
    cache: Arc<Cacher<FetchFileRequest>>,
) -> Box<Future<Item = PrioritizedDownloads, Error = ObjectError>> {
    let index_url = {
        let mut url = source.url.clone();
        if let Some(ref debug_id) = object_id.debug_id {
            url.query_pairs_mut()
                .append_pair("debug_id", &debug_id.to_string());
        }

        if let Some(ref code_id) = object_id.code_id {
            url.query_pairs_mut()
                .append_pair("code_id", &code_id.to_string());
        }

        url
    };

    log::debug!("Fetching list of Sentry debug files from {}", index_url);
    let index_request = clone!(source, || {
        client::get(&index_url)
            .header("User-Agent", USER_AGENT)
            .header("Authorization", format!("Bearer {}", source.token.clone()))
            .finish()
            .unwrap()
            .send()
            .map_err(|e| e.context(SentryErrorKind::SendRequest).into())
            .and_then(move |response| {
                if response.status().is_success() {
                    log::trace!("Success fetching index from Sentry");
                    Either::A(
                        response
                            .json::<Vec<FileEntry>>()
                            .map_err(|e| e.context(SentryErrorKind::Parsing).into()),
                    )
                } else {
                    log::warn!("Sentry returned status code {}", response.status());
                    Either::B(Err(SentryError::from(SentryErrorKind::BadStatusCode)).into_future())
                }
            })
    });

    let index_request = Retry::spawn(
        ExponentialBackoff::from_millis(100).map(jitter).take(3),
        index_request,
    );

    let index_request = future_metrics!("downloads.sentry.index", None, index_request);

    Box::new(
        index_request
            .and_then(clone!(source, object_id, |entries| future::join_all(
                entries.into_iter().map(move |api_response| cache
                    .compute_memoized(FetchFileRequest {
                        scope: scope.clone(),
                        file_id: FileId::Sentry(source.clone(), SentryFileId(api_response.id),),
                        threadpool: threadpool.clone(),
                        object_id: object_id.clone(),
                    })
                    .sentry_hub_new_from_current() // new hub because of join_all
                    .map_err(|e| ArcFail(e).context(ObjectErrorKind::Caching).into())
                    .then(Ok))
            )))
            .map_err(|e| match e {
                tokio_retry::Error::OperationError(e) => e.context(ObjectErrorKind::Sentry).into(),
                e => panic!("{}", e),
            }),
    )
}

pub fn download_from_source(
    source: Arc<SentrySourceConfig>,
    file_id: &SentryFileId,
) -> Box<Future<Item = Option<DownloadStream>, Error = ObjectError>> {
    let download_url = {
        let mut url = source.url.clone();
        url.query_pairs_mut().append_pair("id", &file_id.0);
        url
    };

    log::debug!("Fetching debug file from {}", download_url);
    let token = &source.token;
    let response = clone!(token, download_url, || {
        client::get(&download_url)
            .header("User-Agent", USER_AGENT)
            .header("Authorization", format!("Bearer {}", token))
            // This timeout is for the entire HTTP download *including* the response stream
            // itself, in contrast to what the Actix-Web docs say. We have tested this
            // manually.
            //
            // The intent is to disable the timeout entirely, but there is no API for that.
            .timeout(Duration::from_secs(9999))
            .finish()
            .unwrap()
            .send()
    });

    let response = Retry::spawn(
        ExponentialBackoff::from_millis(100).map(jitter).take(3),
        response,
    );

    let response = response.map_err(|e| match e {
        tokio_retry::Error::OperationError(e) => e,
        e => panic!("{}", e),
    });

    let response = response.then(move |result| match result {
        Ok(response) => {
            if response.status().is_success() {
                log::trace!("Success hitting {}", download_url);
                Ok(Some(Box::new(
                    response
                        .payload()
                        .map_err(|e| e.context(ObjectErrorKind::Io).into()),
                )
                    as Box<dyn Stream<Item = _, Error = _>>))
            } else {
                log::debug!(
                    "Unexpected status code from {}: {}",
                    download_url,
                    response.status()
                );
                Ok(None)
            }
        }
        Err(e) => {
            log::warn!("Skipping response from {}: {}", download_url, e);
            Ok(None)
        }
    });

    Box::new(response)
}
