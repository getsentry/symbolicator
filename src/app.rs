//! Exposes the command line application.
use std::sync::Arc;

use actix_web::App;
use anyhow::{Context, Result};

use crate::actors::{
    cficaches::CfiCacheActor, objects::ObjectsActor, symbolication::SymbolicationActor,
    symcaches::SymCacheActor,
};
use crate::cache::Caches;
use crate::config::Config;
use crate::services::download::DownloadService;
use crate::utils::futures::ThreadPool;

/// The shared state for the service.
#[derive(Clone, Debug)]
pub struct ServiceState {
    /// Thread pool instance reserved for CPU-intensive tasks.
    cpu_threadpool: ThreadPool,
    /// Thread pool instance reserved for IO-intensive tasks.
    io_threadpool: ThreadPool,
    /// Actor for minidump and stacktrace processing
    symbolication: SymbolicationActor,
    /// Actor for downloading and caching objects (no symcaches or cficaches)
    objects: ObjectsActor,
    /// The config object.
    config: Arc<Config>,
    /// The download service.
    download_svc: Arc<DownloadService>,
}

impl ServiceState {
    pub fn create(config: Config) -> Result<Self> {
        let config = Arc::new(config);

        let cpu_threadpool = ThreadPool::new();
        let io_threadpool = ThreadPool::new();

        let download_svc = DownloadService::new(config.clone());

        let caches = Caches::from_config(&config).context("failed to create local caches")?;
        caches
            .clear_tmp(&config)
            .context("failed to clear tmp caches")?;
        let objects = ObjectsActor::new(
            caches.object_meta,
            caches.objects,
            io_threadpool.clone(),
            download_svc.clone(),
        );
        let symcaches =
            SymCacheActor::new(caches.symcaches, objects.clone(), cpu_threadpool.clone());
        let cficaches =
            CfiCacheActor::new(caches.cficaches, objects.clone(), cpu_threadpool.clone());
        let spawnpool = procspawn::Pool::new(config.processing_pool_size)
            .context("failed to create process pool")?;

        let symbolication = SymbolicationActor::new(
            objects.clone(),
            symcaches,
            cficaches,
            caches.diagnostics,
            cpu_threadpool.clone(),
            spawnpool,
        );

        Ok(Self {
            cpu_threadpool,
            io_threadpool,
            symbolication,
            objects,
            config,
            download_svc,
        })
    }

    pub fn io_pool(&self) -> ThreadPool {
        self.io_threadpool.clone()
    }

    pub fn symbolication(&self) -> SymbolicationActor {
        self.symbolication.clone()
    }

    pub fn objects(&self) -> ObjectsActor {
        self.objects.clone()
    }

    pub fn config(&self) -> Arc<Config> {
        self.config.clone()
    }
}

/// Typedef for the application type.
pub type ServiceApp = App<ServiceState>;
