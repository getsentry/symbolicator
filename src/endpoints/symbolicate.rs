use actix_web::{error, App, Error, Json, Query, State};
use serde::Deserialize;

use crate::services::symbolication::SymbolicateStacktraces;
use crate::services::Service;
use crate::sources::SourceConfig;
use crate::types::{
    RawObjectInfo, RawStacktrace, RequestOptions, Scope, Signal, SymbolicationResponse,
};
use crate::utils::sentry::ConfigureScope;

/// Query parameters of the symbolication request.
#[derive(Deserialize)]
pub struct SymbolicationRequestQueryParams {
    #[serde(default)]
    pub timeout: Option<u64>,
    #[serde(default)]
    pub scope: Scope,
}

impl ConfigureScope for SymbolicationRequestQueryParams {
    fn to_scope(&self, scope: &mut sentry::Scope) {
        scope.set_tag("request.scope", &self.scope);
        if let Some(timeout) = self.timeout {
            scope.set_tag("request.timeout", timeout);
        } else {
            scope.set_tag("request.timeout", "none");
        }
    }
}

/// JSON body of the symbolication request.
#[derive(Deserialize)]
struct SymbolicationRequestBody {
    #[serde(default)]
    pub signal: Option<Signal>,
    #[serde(default)]
    pub sources: Option<Vec<SourceConfig>>,
    #[serde(default)]
    pub stacktraces: Vec<RawStacktrace>,
    #[serde(default)]
    pub modules: Vec<RawObjectInfo>,
    #[serde(default)]
    pub options: RequestOptions,
}

async fn symbolicate_frames(
    state: State<Service>,
    params: Query<SymbolicationRequestQueryParams>,
    body: Json<SymbolicationRequestBody>,
) -> Result<Json<SymbolicationResponse>, Error> {
    sentry::start_session();

    let params = params.into_inner();
    params.configure_scope();

    let body = body.into_inner();
    let sources = match body.sources {
        Some(sources) => sources.into(),
        None => state.config().default_sources(),
    };

    let symbolication = state.symbolication();
    let request_id = symbolication.symbolicate_stacktraces(SymbolicateStacktraces {
        scope: params.scope,
        signal: body.signal,
        sources,
        stacktraces: body.stacktraces,
        modules: body.modules.into_iter().map(From::from).collect(),
        options: body.options,
    });

    match symbolication.get_response(request_id, params.timeout).await {
        Some(response) => Ok(Json(response)),
        None => Err(error::ErrorInternalServerError(
            "symbolication request did not start",
        )),
    }
}

pub fn configure(app: App<Service>) -> App<Service> {
    app.resource("/symbolicate", |r| {
        r.post().with_async_config(
            compat_handler!(symbolicate_frames, s, p, b),
            |(_hub, _state, _params, body)| {
                body.limit(5_000_000);
            },
        );
    })
}
