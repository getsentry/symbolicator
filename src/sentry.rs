use std::sync::Arc;

use actix::fut::ActorFuture;
use actix::{Actor, Message};
use futures::future::Future;
use futures::Poll;

use sentry::Hub;

pub struct SentryFuture<F> {
    pub(crate) hub: Arc<Hub>,
    pub(crate) inner: F,
}

impl<F> Future for SentryFuture<F>
where
    F: Future,
{
    type Item = F::Item;
    type Error = F::Error;

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        Hub::run(self.hub.clone(), || self.inner.poll())
    }
}

impl<F> ActorFuture for SentryFuture<F>
where
    F: ActorFuture,
{
    type Actor = F::Actor;
    type Item = F::Item;
    type Error = F::Error;

    fn poll(
        &mut self,
        srv: &mut Self::Actor,
        ctx: &mut <Self::Actor as Actor>::Context,
    ) -> Poll<Self::Item, Self::Error> {
        Hub::run(self.hub.clone(), || self.inner.poll(srv, ctx))
    }
}

pub trait SentryFutureExt: Sized {
    fn sentry_hub(self, hub: Arc<Hub>) -> SentryFuture<Self> {
        SentryFuture { inner: self, hub }
    }

    fn sentry_hub_current(self) -> SentryFuture<Self> {
        self.sentry_hub(Hub::current())
    }

    fn sentry_hub_new_from_current(self) -> SentryFuture<Self> {
        self.sentry_hub(Arc::new(Hub::new_from_top(Hub::current())))
    }
}

// In order to get the extension trait for both actor futures and regular futures, we just
// implement it for everything. If you call `sentry_hub` on random non-future things you just get a
// useless type.
impl<F> SentryFutureExt for F {}

impl<M> Message for SentryFuture<M>
where
    M: Message,
{
    type Result = M::Result;
}

#[macro_export]
macro_rules! handle_sentry_actix_message {
    ($actor:ident, $message:ident) => { handle_sentry_actix_message!(<>, $actor <>, $message <>); };

    (
        < $($param:ident : $param_req:ident),* >,
        $actor:ident < $($param2:ident),* >,
        $message:ident < $($param3:ident),* >
    ) => {
        impl<$($param : $param_req),*> actix::Handler<crate::sentry::SentryFuture<$message<$($param3),*>>> for $actor <$($param2),*> {
            type Result = <$actor <$($param2),*> as actix::Handler<$message<$($param3),*>>>::Result;

            fn handle(&mut self, request: crate::sentry::SentryFuture<$message<$($param3),*>>, ctx: &mut Self::Context) -> Self::Result {
                ::sentry::Hub::run(request.hub.clone(), || self.handle(request.inner, ctx))
            }
        }
    }
}
