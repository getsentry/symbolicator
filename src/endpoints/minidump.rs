use actix_web::{
    dev::Payload, error, http::Method, multipart, Error, HttpMessage, HttpRequest, Json, Query,
    State,
};
use bytes::Bytes;
use futures::{FutureExt, TryFutureExt};
use futures01::{future, Future, Stream};
use sentry::{configure_scope, Hub};

use crate::actors::symbolication::SymbolicationActor;
use crate::app::{ServiceApp, ServiceState};
use crate::endpoints::symbolicate::SymbolicationRequestQueryParams;
use crate::sources::SourceConfig;
use crate::types::{RequestId, RequestOptions, Scope, SymbolicationResponse};
use crate::utils::futures::{ResponseFuture, ThreadPool};
use crate::utils::multipart::{
    read_multipart_file, read_multipart_request_options, read_multipart_sources,
};
use crate::utils::sentry::{ActixWebHubExt, SentryFutureExt, WriteSentryScope};

#[derive(Debug, Default)]
struct MinidumpRequest {
    sources: Option<Vec<SourceConfig>>,
    minidump: Option<Bytes>,
    options: RequestOptions,
}

fn handle_multipart_item(
    threadpool: ThreadPool,
    mut request: MinidumpRequest,
    item: multipart::MultipartItem<Payload>,
) -> ResponseFuture<MinidumpRequest, Error> {
    let field = match item {
        multipart::MultipartItem::Field(field) => field,
        multipart::MultipartItem::Nested(nested) => {
            return handle_multipart_stream(threadpool, request, nested);
        }
    };

    match field
        .content_disposition()
        .as_ref()
        .and_then(|d| d.get_name())
    {
        Some("sources") => {
            let future = read_multipart_sources(field).map(move |sources| {
                request.sources = Some(sources);
                request
            });
            Box::new(future)
        }
        Some("upload_file_minidump") => {
            let future = read_multipart_file(field).map(move |minidump| {
                request.minidump = Some(minidump);
                request
            });
            Box::new(future)
        }
        Some("options") => {
            let future = read_multipart_request_options(field).map(move |options| {
                request.options = options;
                request
            });
            Box::new(future)
        }
        _ => {
            // Always ignore unknown fields.
            Box::new(future::ok(request))
        }
    }
}

fn handle_multipart_stream(
    threadpool: ThreadPool,
    request: MinidumpRequest,
    stream: multipart::Multipart<Payload>,
) -> ResponseFuture<MinidumpRequest, Error> {
    let future = stream
        .map_err(Error::from)
        .fold(request, move |request, item| {
            handle_multipart_item(threadpool.clone(), request, item)
        });

    Box::new(future)
}

fn process_minidump(
    symbolication: &SymbolicationActor,
    request: MinidumpRequest,
    scope: Scope,
) -> Result<RequestId, Error> {
    let minidump = request
        .minidump
        .ok_or_else(|| error::ErrorBadRequest("missing minidump"))?;

    sentry::configure_scope(|scope| {
        let mut hasher = crc32fast::Hasher::new();
        hasher.update(&minidump);
        scope.set_extra("minidump_crc32", hasher.finalize().into());
        scope.set_extra("minidump_len", minidump.len().into());
    });

    let sources = request
        .sources
        .ok_or_else(|| error::ErrorBadRequest("missing sources"))?;

    Ok(symbolication.process_minidump(scope, minidump, sources, request.options))
}

fn handle_minidump_request(
    state: State<ServiceState>,
    params: Query<SymbolicationRequestQueryParams>,
    request: HttpRequest<ServiceState>,
) -> ResponseFuture<Json<SymbolicationResponse>, Error> {
    let hub = Hub::from_request(&request);
    hub.start_session();

    Hub::run(hub, || {
        let default_sources = state.config().default_sources();

        let params = params.into_inner();
        configure_scope(|scope| {
            params.write_sentry_scope(scope);
        });

        let request_future = handle_multipart_stream(
            state.io_pool(),
            MinidumpRequest::default(),
            request.multipart(),
        );

        let SymbolicationRequestQueryParams { scope, timeout } = params;
        let symbolication = state.symbolication();

        let response_future = request_future
            .and_then(clone!(symbolication, |mut request| {
                if request.sources.is_none() {
                    request.sources = Some(default_sources.to_vec());
                }

                process_minidump(&symbolication, request, scope)
            }))
            .and_then(move |request_id| {
                symbolication
                    .get_response(request_id, timeout)
                    .never_error()
                    .boxed_local()
                    .compat()
                    .then(|result| match result {
                        Ok(Some(response)) => Ok(Json(response)),
                        Ok(None) => Err(error::ErrorInternalServerError(
                            "symbolication request did not start",
                        )),
                        Err(never) => match never {},
                    })
                    .map_err(Error::from)
            });

        Box::new(response_future.sentry_hub_current())
    })
}

pub fn configure(app: ServiceApp) -> ServiceApp {
    app.resource("/minidump", |r| {
        r.method(Method::POST).with(handle_minidump_request);
    })
}

#[cfg(test)]
mod tests {
    use std::fs;

    use actix_web::test::TestServer;
    use reqwest::{multipart, Client, StatusCode};

    use crate::app::ServiceState;
    use crate::config::Config;
    use crate::test;
    use crate::types::SymbolicationResponse;

    #[tokio::test]
    async fn test_basic() {
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", "[]");

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = response.text().await.unwrap();
        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
        insta::assert_yaml_snapshot!(response);
    }

    // This test is disabled because it locks up on CI. We have not found a way to reproduce this.
    #[allow(dead_code)]
    // #[tokio::test]
    async fn test_integration_microsoft() {
        // TODO: Move this test to E2E tests
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));
        let source = test::microsoft_symsrv();

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", serde_json::to_string(&vec![source]).unwrap())
            .text("options", r#"{"dif_candidates":true}"#);

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = response.text().await.unwrap();
        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
        insta::assert_yaml_snapshot!(response);
    }

    #[tokio::test]
    async fn test_unknown_field() {
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", "[]")
            .text("unknown", "value");

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
}
