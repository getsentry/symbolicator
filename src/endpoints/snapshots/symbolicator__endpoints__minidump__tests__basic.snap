---
source: src/endpoints/minidump.rs
expression: response
---
status: completed
timestamp: 1521713273
system_info:
  os_name: Windows
  os_version: 10.0.14393
  os_build: ""
  cpu_arch: x86
  device_model: ""
crashed: true
crash_reason: EXCEPTION_ACCESS_VIOLATION_WRITE
assertion: ""
stacktraces:
  - thread_id: 1636
    is_requesting: true
    registers:
      eax: "0x0"
      ebp: "0x10ff670"
      ebx: "0xfe5000"
      ecx: "0x10ff670"
      edi: "0x13bfd78"
      edx: "0x7"
      eflags: "0x10246"
      eip: "0x2a2a3d"
      esi: "0x759c6314"
      esp: "0x10ff644"
    frames:
      - status: missing
        original_index: 0
        instruction_addr: "0x2a2a3d"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: context
      - status: missing
        original_index: 1
        instruction_addr: "0x2a28d0"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: fp
      - status: missing
        original_index: 2
        instruction_addr: "0x7584e9c0"
        package: "C:\\Windows\\System32\\rpcrt4.dll"
        trust: scan
      - status: missing
        original_index: 3
        instruction_addr: "0x70b7ae40"
        package: "C:\\Windows\\System32\\dbgcore.dll"
        trust: scan
      - status: missing
        original_index: 4
        instruction_addr: "0x7584e9c0"
        package: "C:\\Windows\\System32\\rpcrt4.dll"
        trust: scan
      - status: missing
        original_index: 5
        instruction_addr: "0x2a3435"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: scan
      - status: missing
        original_index: 6
        instruction_addr: "0x2a2d97"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: scan
      - status: missing
        original_index: 7
        instruction_addr: "0x750662c4"
        package: "C:\\Windows\\System32\\kernel32.dll"
        trust: fp
      - status: missing
        original_index: 8
        instruction_addr: "0x771d0f79"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
      - status: missing
        original_index: 9
        instruction_addr: "0x771d0f44"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
  - thread_id: 3580
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x159faa4"
      ebx: "0x13b0990"
      ecx: "0x0"
      edi: "0x13b4af0"
      edx: "0x0"
      eflags: "0x216"
      eip: "0x771e016c"
      esi: "0x13b4930"
      esp: "0x159f900"
    frames:
      - status: missing
        original_index: 0
        instruction_addr: "0x771e016c"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: context
      - status: missing
        original_index: 1
        instruction_addr: "0x750662c4"
        package: "C:\\Windows\\System32\\kernel32.dll"
        trust: fp
      - status: missing
        original_index: 2
        instruction_addr: "0x771d0f79"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
      - status: missing
        original_index: 3
        instruction_addr: "0x771d0f44"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
  - thread_id: 2600
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x169fb98"
      ebx: "0x13b0990"
      ecx: "0x0"
      edi: "0x13b7c28"
      edx: "0x0"
      eflags: "0x202"
      eip: "0x771e016c"
      esi: "0x13b7a68"
      esp: "0x169f9f4"
    frames:
      - status: missing
        original_index: 0
        instruction_addr: "0x771e016c"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: context
      - status: missing
        original_index: 1
        instruction_addr: "0x750662c4"
        package: "C:\\Windows\\System32\\kernel32.dll"
        trust: fp
      - status: missing
        original_index: 2
        instruction_addr: "0x771d0f79"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
      - status: missing
        original_index: 3
        instruction_addr: "0x771d0f44"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: fp
  - thread_id: 2920
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x179f2b8"
      ebx: "0x17b1aa0"
      ecx: "0x0"
      edi: "0x17b1a90"
      edx: "0x0"
      eflags: "0x206"
      eip: "0x771df3dc"
      esi: "0x2cc"
      esp: "0x179f2ac"
    frames:
      - status: missing
        original_index: 0
        instruction_addr: "0x771df3dc"
        package: "C:\\Windows\\System32\\ntdll.dll"
        trust: context
modules:
  - debug_status: missing
    unwind_status: missing
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 5ab380779000
    code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
    debug_id: 3249d99d-0c40-4931-8610-f4e4fb0b6936-1
    debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
    image_addr: "0x2a0000"
    image_size: 36864
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: "57898e12145000"
    code_file: "C:\\Windows\\System32\\dbghelp.dll"
    debug_id: 9c2a902b-6fdf-40ad-8308-588a41d572a0-1
    debug_file: dbghelp.pdb
    image_addr: "0x70850000"
    image_size: 1331200
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 589abc846c000
    code_file: "C:\\Windows\\System32\\msvcp140.dll"
    debug_id: bf5257f7-8c26-43dd-9bb7-901625e1136a-1
    debug_file: msvcp140.i386.pdb
    image_addr: "0x709a0000"
    image_size: 442368
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57898eeb92000
    code_file: "C:\\Windows\\System32\\apphelp.dll"
    debug_id: 8daf7773-372f-460a-af38-944e193f7e33-1
    debug_file: apphelp.pdb
    image_addr: "0x70a10000"
    image_size: 598016
  - debug_status: missing
    unwind_status: missing
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57898dab25000
    code_file: "C:\\Windows\\System32\\dbgcore.dll"
    debug_id: aec7ef2f-df4b-4642-a471-4c3e5fe8760a-1
    debug_file: dbgcore.pdb
    image_addr: "0x70b70000"
    image_size: 151552
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 589abc7714000
    code_file: "C:\\Windows\\System32\\VCRUNTIME140.dll"
    debug_id: 0ed80a50-ecda-472b-86a4-eb6c833f8e1b-1
    debug_file: vcruntime140.i386.pdb
    image_addr: "0x70c60000"
    image_size: 81920
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57899141a000
    code_file: "C:\\Windows\\System32\\CRYPTBASE.dll"
    debug_id: 147c51fb-7ca1-408f-85b5-285f2ad6f9c5-1
    debug_file: cryptbase.pdb
    image_addr: "0x73ba0000"
    image_size: 40960
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59bf30e31f000
    code_file: "C:\\Windows\\System32\\sspicli.dll"
    debug_id: 51e432b1-0450-4b19-8ed1-6d4335f9f543-1
    debug_file: wsspicli.pdb
    image_addr: "0x73bb0000"
    image_size: 126976
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 5a49bb7677000
    code_file: "C:\\Windows\\System32\\advapi32.dll"
    debug_id: 0c799483-b549-417d-8433-4331852031fe-1
    debug_file: advapi32.pdb
    image_addr: "0x73c70000"
    image_size: 487424
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57899155be000
    code_file: "C:\\Windows\\System32\\msvcrt.dll"
    debug_id: 6f6409b3-d520-43c7-9b2f-62e00bfe761c-1
    debug_file: msvcrt.pdb
    image_addr: "0x73cf0000"
    image_size: 778240
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 598942c741000
    code_file: "C:\\Windows\\System32\\sechost.dll"
    debug_id: 6f6a05dd-0a80-478b-a419-9b88703bf75b-1
    debug_file: sechost.pdb
    image_addr: "0x74450000"
    image_size: 266240
  - debug_status: missing
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 590285e9e0000
    code_file: "C:\\Windows\\System32\\kernel32.dll"
    debug_id: d3474559-96f7-47d6-bf43-c176b2171e68-1
    debug_file: wkernel32.pdb
    image_addr: "0x75050000"
    image_size: 917504
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59b0df8f5a000
    code_file: "C:\\Windows\\System32\\bcryptPrimitives.dll"
    debug_id: 287b19c3-9209-4a2b-bb8f-bcc37f411b11-1
    debug_file: bcryptprimitives.pdb
    image_addr: "0x75130000"
    image_size: 368640
  - debug_status: missing
    unwind_status: missing
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 5a49bb75c1000
    code_file: "C:\\Windows\\System32\\rpcrt4.dll"
    debug_id: ae131c67-27a7-4fa1-9916-b5a4aef41190-1
    debug_file: wrpcrt4.pdb
    image_addr: "0x75810000"
    image_size: 790528
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59bf2b5ae0000
    code_file: "C:\\Windows\\System32\\ucrtbase.dll"
    debug_id: 6bedcbce-0a3a-40e9-8040-81c2c8c6cc2f-1
    debug_file: ucrtbase.pdb
    image_addr: "0x758f0000"
    image_size: 917504
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59bf2bcf1a1000
    code_file: "C:\\Windows\\System32\\KERNELBASE.dll"
    debug_id: 8462294a-c645-402d-ac82-a4e95f61ddf9-1
    debug_file: wkernelbase.pdb
    image_addr: "0x76db0000"
    image_size: 1708032
  - debug_status: missing
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59b0d8f3183000
    code_file: "C:\\Windows\\System32\\ntdll.dll"
    debug_id: 971f98e5-ce60-41ff-b2d7-235bbeb34578-1
    debug_file: wntdll.pdb
    image_addr: "0x77170000"
    image_size: 1585152
